### ISSUE TITLE
Add type hints to signatures of basic linker types

### ISSUE THREAD

**Issue by michaelosthege** [pymc-devs/aesara#229]
After #228 is merged - or even on the same branch we urgently need type <s>hints</s> <s>annotations</s> hints on some core classes:
+ `Linker`, `LocalLinker`, `PerformLinker`
+ `VM`
+ `FunctionGraph`

And also on the related helper functions such as `streamline` or `map_storage`.

The description under this links is a good entry point for starters: https://theano-pymc.readthedocs.io/en/latest/extending/pipeline.html

**Comment by LegrandNico** [comment]
Should we use type annotation or type hints here? The title refers to annotations but the text mention type hints.

I am asking because type hints would probably require to add several class imports, and probability some circular ones, for which fixes do exist but not for all Python versions (e.g https://docs.python.org/3.7/whatsnew/3.7.html#pep-563-postponed-evaluation-of-annotations).

For example, I am using [MonkeyType ](https://github.com/Instagram/MonkeyType) to add type hints for the `theano.link.basic`submodule, which works correctly but requires e.g. to import `from theano.graph.utils.MetaObject` for the `Container` class.

How should we deal with such external class imports? Just add it to the import list? Should we use the `typing.TYPE_CHECKING` condition to ensure it is not executed at runtime?

**Comment by michaelosthege** [comment]
Sorry, I was not aware of a difference between hint and annotations. I meant type annotations like `name : str`.

We should not introduce circular imports. If a type annotation is not possible because of import dependencies, then we'll have to skip it and resolve the import dependency first.

**Comment by LegrandNico** [comment]
For the type hints/annotation distinction, I am referring to this [post](https://testdriven.io/blog/python-type-checking/#Type-Annotations-vs-Type-Hints) (I don't know if this is still up to date), but accordingly using `name: str` would rather be a type hint, as it can be checked for consistency using e.g. [mypy](https://mypy.readthedocs.io/en/stable/).

Also, I noted that when checking for type consistencies, running `mypy ./theano/ --ignore-missing-imports --follow-imports=skip` returns several inconsistencies in types assignments in the codebase. It is something that should be fixed too, so in the future, mypy could be used for static checks in CI (and maybe in the pre-commit hooks)?

**Comment by brandonwillard** [comment]
> Is it something that should be fixed too, so in the future, mypy could be used for static checks in CI (and maybe in the pre-commit hooks)?

Absolutely
